File: app/src/main/java/com/example/my_hospital_appointments/DoctorAppointments.java
package com.example.my_hospital_appointments;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.RelativeLayout;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;

public class DoctorAppointments extends AppCompatActivity {

    private ArrayList<String> myArrayList=new ArrayList<String>();
    DatabaseReference myDoctorsData;
    DatabaseReference attendedAppointmentsData;
    Spinner spinnerEmails;
    public String patientEmail="";
    public TextView textViewPatientName, textViewPatientDescription, textViewPatientDate, textViewPatientTime;
    Button btnViewAppointments,btnCancelAppointments,btnAttendAppointments,backHomeAppointments;
    private String patientName,patientDescription,patientDate,patientTime;
    RelativeLayout myLayoutAttend;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_doctor_appointments);

        myLayoutAttend=(RelativeLayout)  findViewById(R.id.relativeLayoutViewAttend);
        attendedAppointmentsData=FirebaseDatabase.getInstance().getReference("AttendedAppointments");

        textViewPatientName =(TextView)  findViewById(R.id.textViewPatientNameDoctorAppointments2);
        textViewPatientDescription =(TextView)  findViewById(R.id.textViewPatientDescriptionDoctorAppointments2);
        textViewPatientDate =(TextView)  findViewById(R.id.textViewPatientDateDoctorAppointments2);
        textViewPatientTime =(TextView)  findViewById(R.id.textViewPatientTimeDoctorAppointments2);

        Intent intent =getIntent();
        String myUsersEmail=intent.getExtras().getString("usersEmail");

        spinnerEmails =(Spinner) findViewById(R.id.spinnerDoctorAppointments);

        myArrayList.add("Select One");

        myDoctorsData = FirebaseDatabase.getInstance().getReference().child("AssignedPatient");
        myDoctorsData.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                for(DataSnapshot snapshot : dataSnapshot.getChildren())
                {
                    myArrayList.add(snapshot.child("email").getValue().toString());
                }

            }
            @Override
            public void onCancelled(@NonNull DatabaseError error) {

            }
        });


        ArrayAdapter adapter1 = new ArrayAdapter(this, android.R.layout.simple_spinner_item,myArrayList);
        adapter1.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);

        spinnerEmails.setAdapter(adapter1);
        spinnerEmails.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {

                String myText=spinnerEmails.getSelectedItem().toString().trim();
                if(myText=="Select One".trim())
                {
                    //Nothing

                }
                else
                {
                    Toast.makeText(DoctorAppointments.this,"You have Selected "+myText,Toast.LENGTH_SHORT).show();
                    patientEmail=myText;
                }
            }
            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });

        btnViewAppointments=(Button) findViewById(R.id.buttonDoctorViewAppointments);
        btnViewAppointments.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                ViewAppointment(myUsersEmail);
            }
        });

        btnAttendAppointments=(Button) findViewById(R.id.buttonApointmentAttendedDoctor);
        btnAttendAppointments.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                AttendedAppointments(myUsersEmail);
            }
        });

        btnCancelAppointments=(Button) findViewById(R.id.buttonCancelAppointmentDoctor);
        btnCancelAppointments.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                myDoctorsData.child(myUsersEmail).removeValue();
                Toast.makeText(DoctorAppointments.this,"APPOINTMENT CANCELLED SUCCESSSFULLY",Toast.LENGTH_SHORT).show();
            }
        });

    }

    public void ViewAppointment(String doctorEmail)
    {

        String emailKey="";
        int Counter=doctorEmail.length();

        for(int a=0; a<Counter; a++)
        {
            if(doctorEmail.charAt(a)=='@')
            {
                break;
            }
            else
            {
                emailKey=emailKey+doctorEmail.charAt(a);

            }
        }

        emailKey=emailKey.trim();

        myDoctorsData.child(emailKey).get().addOnCompleteListener(new OnCompleteListener<DataSnapshot>() {
            @Override
            public void onComplete(@NonNull Task<DataSnapshot> task) {

                if(task.isSuccessful())
                {
                    if(task.getResult().exists())
                    {
                        // public TextView patientName,patientDescription,patientDate,patientTime;
                        DataSnapshot thisDataSnapshot=task.getResult();
                         patientName=String.valueOf(thisDataSnapshot.child("name").getValue());
                         patientDescription=String.valueOf(thisDataSnapshot.child("description").getValue());
                         patientDate=String.valueOf(thisDataSnapshot.child("date").getValue());
                         patientTime=String.valueOf(thisDataSnapshot.child("time").getValue());

                        textViewPatientName.setText(patientName);
                        textViewPatientDate.setText(patientDate);
                        textViewPatientDescription.setText(patientDescription);
                        textViewPatientTime.setText(patientTime);
                    }
                }
            }
        });

    }

    public void AttendedAppointments(String doctorEmail)
    {

        String emailKey="";
        int Counter=doctorEmail.length();

        for(int a=0; a<Counter; a++)
        {
            if(doctorEmail.charAt(a)=='@')
            {
                break;
            }
            else
            {
                emailKey=emailKey+doctorEmail.charAt(a);

            }
        }

        emailKey=emailKey.trim();

        //String id=databaseAppointments.push().getKey(); //unique path that identifies the location of the data
        String id=attendedAppointmentsData.push().getKey();
        //patientName,patientDescription,patientDate,patientTime
        //String patientEmail, String doctorEmail, String description, String time, String date
        AppointmentsAttended myAppointments=new AppointmentsAttended(patientEmail,doctorEmail,patientDescription,patientTime,patientDate);
        String finalEmailKey = emailKey;
        attendedAppointmentsData.child(id).setValue(myAppointments).addOnCompleteListener(new OnCompleteListener<Void>() {
            @Override
            public void onComplete(@NonNull Task<Void> task) {
                if(task.isSuccessful())
                {

                    myDoctorsData.child(finalEmailKey).removeValue();
                    btnViewAppointments.setVisibility(View.GONE);
                    btnAttendAppointments.setVisibility(View.GONE);
                    btnCancelAppointments.setVisibility(View.GONE);
                    myLayoutAttend.setVisibility(View.VISIBLE);
                    backHomeAppointments=(Button) findViewById(R.id.buttonBackAfterAppointments);
                    backHomeAppointments.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View view) {
                            Intent myIntent=new Intent(DoctorAppointments.this,MainPageDoctor.class);
                            startActivity(myIntent);
                        }
                    });
                }
            }
        });
    }
}

File: app/src/main/java/com/example/my_hospital_appointments/DoctorAppointments.kt
package com.example.my_hospital_appointments

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import com.google.android.gms.tasks.OnCompleteListener
import com.google.android.gms.tasks.Task
import com.google.firebase.database.*

class DoctorAppointments() : AppCompatActivity() {
    private val myArrayList = ArrayList<String?>()
    var myDoctorsData: DatabaseReference? = null
    var attendedAppointmentsData: DatabaseReference? = null
    var spinnerEmails: Spinner? = null
    var patientEmail = ""
    var textViewPatientName: TextView? = null
    var textViewPatientDescription: TextView? = null
    var textViewPatientDate: TextView? = null
    var textViewPatientTime: TextView? = null
    var btnViewAppointments: Button? = null
    var btnCancelAppointments: Button? = null
    var btnAttendAppointments: Button? = null
    var backHomeAppointments: Button? = null
    private var patientName: String? = null
    private var patientDescription: String? = null
    private var patientDate: String? = null
    private var patientTime: String? = null
    var myLayoutAttend: RelativeLayout? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_doctor_appointments)
        myLayoutAttend = findViewById<View>(R.id.relativeLayoutViewAttend) as RelativeLayout
        attendedAppointmentsData =
            FirebaseDatabase.getInstance().getReference("AttendedAppointments")
        textViewPatientName =
            findViewById<View>(R.id.textViewPatientNameDoctorAppointments2) as TextView
        textViewPatientDescription =
            findViewById<View>(R.id.textViewPatientDescriptionDoctorAppointments2) as TextView
        textViewPatientDate =
            findViewById<View>(R.id.textViewPatientDateDoctorAppointments2) as TextView
        textViewPatientTime =
            findViewById<View>(R.id.textViewPatientTimeDoctorAppointments2) as TextView
        val intent = intent
        val myUsersEmail = intent.extras!!.getString("usersEmail")
        spinnerEmails = findViewById<View>(R.id.spinnerDoctorAppointments) as Spinner
        myArrayList.add("Select One")
        myDoctorsData = FirebaseDatabase.getInstance().reference.child("AssignedPatient")
        myDoctorsData!!.addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                for (snapshot: DataSnapshot in dataSnapshot.children) {
                    myArrayList.add(snapshot.child("email").value.toString())
                }
            }

            override fun onCancelled(error: DatabaseError) {}
        })
        val adapter1: ArrayAdapter<*> =
            ArrayAdapter<Any?>(this, android.R.layout.simple_spinner_item,
                myArrayList as List<Any?>
            )
        adapter1.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        spinnerEmails!!.adapter = adapter1
        spinnerEmails!!.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>?, view: View, i: Int, l: Long) {
                val myText: String = spinnerEmails!!.selectedItem.toString().trim { it <= ' ' }
                if (myText === "Select One".trim { it <= ' ' }) {
                    //Nothing
                } else {
                    Toast.makeText(
                        this@DoctorAppointments,
                        "You have Selected $myText",
                        Toast.LENGTH_SHORT
                    ).show()
                    patientEmail = myText
                }
            }

            override fun onNothingSelected(adapterView: AdapterView<*>?) {}
        }
        btnViewAppointments = findViewById<View>(R.id.buttonDoctorViewAppointments) as Button
        btnViewAppointments!!.setOnClickListener(View.OnClickListener { ViewAppointment(myUsersEmail) })
        btnAttendAppointments = findViewById<View>(R.id.buttonApointmentAttendedDoctor) as Button
        btnAttendAppointments!!.setOnClickListener(object : View.OnClickListener {
            override fun onClick(view: View) {
                AttendedAppointments(myUsersEmail)
            }
        })
        btnCancelAppointments = findViewById<View>(R.id.buttonCancelAppointmentDoctor) as Button
        btnCancelAppointments!!.setOnClickListener(object : View.OnClickListener {
            override fun onClick(view: View) {
                myDoctorsData!!.child((myUsersEmail)!!).removeValue()
                Toast.makeText(
                    this@DoctorAppointments,
                    "APPOINTMENT CANCELLED SUCCESSSFULLY",
                    Toast.LENGTH_SHORT
                ).show()
            }
        })
    }

    fun ViewAppointment(doctorEmail: String?) {
        var emailKey = ""
        val Counter = doctorEmail!!.length
        for (a in 0 until Counter) {
            if (doctorEmail[a] == '@') {
                break
            } else {
                emailKey = emailKey + doctorEmail[a]
            }
        }
        emailKey = emailKey.trim { it <= ' ' }
        myDoctorsData!!.child(emailKey).get()
            .addOnCompleteListener(object : OnCompleteListener<DataSnapshot> {
                override fun onComplete(task: Task<DataSnapshot>) {
                    if (task.isSuccessful) {
                        if (task.result.exists()) {
                            // public TextView patientName,patientDescription,patientDate,patientTime;
                            val thisDataSnapshot = task.result
                            patientName = thisDataSnapshot.child("name").value.toString()
                            patientDescription =
                                thisDataSnapshot.child("description").value.toString()
                            patientDate = thisDataSnapshot.child("date").value.toString()
                            patientTime = thisDataSnapshot.child("time").value.toString()
                            textViewPatientName!!.text = patientName
                            textViewPatientDate!!.text = patientDate
                            textViewPatientDescription!!.text = patientDescription
                            textViewPatientTime!!.text = patientTime
                        }
                    }
                }
            })
    }

    fun AttendedAppointments(doctorEmail: String?) {
        var emailKey = ""
        val Counter = doctorEmail!!.length
        for (a in 0 until Counter) {
            if (doctorEmail[a] == '@') {
                break
            } else {
                emailKey = emailKey + doctorEmail[a]
            }
        }
        emailKey = emailKey.trim { it <= ' ' }

        //String id=databaseAppointments.push().getKey(); //unique path that identifies the location of the data
        val id = attendedAppointmentsData!!.push().key
        //patientName,patientDescription,patientDate,patientTime
        //String patientEmail, String doctorEmail, String description, String time, String date
        val myAppointments = AppointmentsAttended(
            patientEmail,
            doctorEmail,
            patientDescription,
            patientTime,
            patientDate
        )
        val finalEmailKey = emailKey
        attendedAppointmentsData!!.child((id)!!).setValue(myAppointments)
            .addOnCompleteListener(object : OnCompleteListener<Void?> {
                override fun onComplete(task: Task<Void?>) {
                    if (task.isSuccessful) {
                        myDoctorsData!!.child(finalEmailKey).removeValue()
                        btnViewAppointments!!.visibility = View.GONE
                        btnAttendAppointments!!.visibility = View.GONE
                        btnCancelAppointments!!.visibility = View.GONE
                        myLayoutAttend!!.visibility = View.VISIBLE
                        backHomeAppointments =
                            findViewById<View>(R.id.buttonBackAfterAppointments) as Button
                        backHomeAppointments!!.setOnClickListener(object : View.OnClickListener {
                            override fun onClick(view: View) {
                                val myIntent =
                                    Intent(this@DoctorAppointments, MainPageDoctor::class.java)
                                startActivity(myIntent)
                            }
                        })
                    }
                }
            })
    }
}